### ISSUE TITLE
Helpers are not refreshing in development mode

### ISSUE THREAD

**Issue by railsnerd** [activeadmin/inherited_resources#95]
Hi there

Normal development Rails behavior is for changes to a helper to be automatically reflected on page refresh. A server reload is not necessary.

However this gem needs a server reload for the helper to get updated.

Is there a work around for this?  I hate tested in Rails 3.1

thanks

**Comment by railsnerd** [comment]
OK so I found this ticket https://github.com/josevalim/inherited_resources/issues/closed#issue/81

Is inheriting from inside the controller the official solution?  In the comments this is proposed as a temporary fix?

Would be nice to keep using "class X < InheritedResources::Base"

**Comment by josevalim** [comment]
This is a bug and a patch is welcome.

**Comment by collin** [comment]
I'm not sure how to fix this bug, but maybe this information helps:

Upon booting a rails console ( with X < InheritedResources::Base )

```
  > ActiveSupport::Dependencies.autoloaded_constants
 => ["Ability", "Article", "Publication", "Staff", "User"] 
  > StaffsController
 => StaffsController 
  > ActiveSupport::Dependencies.autoloaded_constants
 => ["Ability", "Article", "Publication", "Staff", "User", "DeviseHelper", "LinkHelper", "ProfileHelper", "ApplicationController", "StaffsController"] 
  > reload!
Reloading...
 => true 
  > StaffsController
 => StaffsController 
  > ActiveSupport::Dependencies.autoloaded_constants
 => ["Ability", "Article", "Publication", "Staff", "User", "StaffsController"]
```

I actually had my first debugging session that involved breaking of constants autoloading. In that case I was using require in a file that was being autoloaded, which short circuited the code in ActiveSupport::Dependencies that would have registered the class for reloading.

The fix I employed there was to use Kernel.require instead of require, which implicitly is called on Object. Rails overrides Object.require, and leaves Kernel.require alone.

I peeked into inherited_resources/base, but that strategy had no effect.

If somebody knows a good resource for which actions will short-circuit ActiveSupport::Dependencies that would be a great thing to have when approaching this bug.

**Comment by albertoperdomo** [comment]
Is this still an issue? I'm seeing this behaviour as well...

**Comment by borisnadion** [comment]
looks like it is

**Comment by marcandre** [comment]
+1

**Comment by escobera** [comment]
Need this too.

**Comment by vitaly** [comment]
hey, just wasted lots of time debugging this, finally traced to inherited_resource. Any ETA on the fix?

**Comment by thedarkone** [comment]
I think this is the same sure I run into with `rails-dev-boost`: https://github.com/thedarkone/rails-dev-boost/issues/8

**Comment by josevalim** [comment]
Is this an issue just when using rails-dev-boost? So is this fixed now?

**Comment by borisnadion** [comment]
no, it happens without rails-dev-boost

**Comment by josevalim** [comment]
Fixed in master. Please check! If it works I will release a new version.

**Comment by codespore** [comment]
It's still happening.

**Comment by thedarkone** [comment]
@josevalim this is best fixed by turning `InheritedResources::Base` into "development mode reloadable" constant via something like this: [https://github.com/thedarkone/inherited_resources/commit/9886f20de211a550ab0d81a98fa11e38676b35c3](https://github.com/thedarkone/inherited_resources/commit/9886f20de211a550ab0d81a98fa11e38676b35c3).

This way there is no need to have any helpers related hacks.

**Comment by collin** [comment]
Oh, hey. Is InheritedResources::Base being autoloaded by the Rails mechanism?

Looks like it is.

I wasn't able to find any documentation on this but here is what I discovered through trial and error/going over the Rails source.

When the rails autoloader is invoked, say by accessing a constant that hasn't  been loaded yet, Rails goes out in search for the proper file to load for that constant.

Rails ordinarily marks this constant as reloadable.

There is a special case ( there may be more, but I am sure of this one ) that causes the constant to NOT be marked as reloadable.

That is, if you call "require" in the file that rails is loading for this constant.

Rails does this by redefining the require method. In the redefinition there is a "short-circuit" for marking classes as reloadable.

There is a way around this. Use "Kernel.require" in files for classes you want to be reloadable. Rails leaves that method alone.

**Comment by thedarkone** [comment]
As it is right now `InheritedResources::Base` is not being auto-reloaded by Rails because it is in gem's `lib` directory. However it does inherit from `ApplicationController` that is an application class, that **is** auto-reloaded on per request basis.

So things look like this:

`ApplicationController` [reloadable] -> `InheritedResources::Base` [non-reloadable] -> `YourAppController` [reloadable].

Once you hit refresh `ApplicationController` and `YourAppController` is unloaded by Rails.  `InheritResources::Base` is now inheriting from an `ApplicationController` class that has been unloaded by Rails and a new `ApplicationController` has been created/redefined.

The new `ApplicationController` is completely different class than the one `InheritedResources::Base` inherits from. Therefore any changes made to `ApplicationController` file are not visible to the classes inheriting from `InheritedResources::Base`.

Some more explanation is also in this ticket: https://github.com/thedarkone/rails-dev-boost/issues/8

**Comment by volkanunsal** [comment]
So what's the fix for this issue? I'm also seeing it in development mode.

**Comment by gamafranco** [comment]
+1

**Comment by thedarkone** [comment]
Fix is an pull request: #145.
